Docs: Split Custom Domains as Explanation and How-to Guide (Diátaxis) (#9676)
readthedocs/readthedocs.org: docs/user/custom-domains.rst

Custom Domains
==============

You can serve your documentation project from your own domain,
for instance ``docs.example.com``.
This is great for maintaining a consistent brand for your product and its documentation.

*By default*, your documentation is served from a Read the Docs :ref:`subdomain <hosting:subdomain support>` using the project's :term:`slug`:

* ``<slug>.readthedocs.io`` for |org_brand|
* ``<slug>.readthedocs-hosted.com`` for |com_brand|.

.. seealso::

    :doc:`/guides/custom-domains`
        Information on creating and managing custom domains,
        and common configurations you might use to set up your domain

How custom domains work
-----------------------

To use a custom domain, you enter the domain in your Read the Docs project's :guilabel:`Admin` and update your DNS provider with a new DNS entry.

These two actions are all that are needed from you.
Once the DNS record has propagated,
Read the Docs automatically issues an SSL certificate through Cloudflare and starts serving your documentation.

.. mermaid::

    graph TD
        subgraph rtd [On Read the Docs]
          A(fa:fa-pencil Add docs.example.com as Custom domain)

>           A-->AA(fa:fa-spinner Your project is rebuilt with<br>docs.example.com as a canonical domain)

Reviewer: This won't happen automatically, will it? They need to rebuild the project after the canonical domain is properly configured.

I think this should be moved to after the `Wait for DNS` step, probably at the end? 

Something like `Next time your project is built, it will update the :doc:/canonical-urls to use docs.example.com`

        end
        subgraph dns [On your domain's DNS administration]

>           B(fa:fa-pencil Edit DNS entry for docs.example.com)

Reviewer: ```suggestion
          B(fa:fa-pencil Edit DNS entry for ``docs.example.com`` to point at the endpoint noted on your Read the Docs domain page.)
```

Good to be explicit here.

Author: I'll have to think of something here that's explicit :+1: 

I don't think the `<code>` part will work, but will see what's possible here.

        end

>         rtd & dns-->C(fa:fa-spinner Wait for DNS propogation<br>Usually just a few minutes)

Reviewer: ```suggestion
        rtd & dns-->C(fa:fa-spinner Wait for DNS propagation.<br>Usually just a few minutes)
```

        direction LR

>         subgraph automatic [The rest is handled automatically]

Reviewer: ```suggestion
        subgraph automatic [fa:fa-sparkles The rest is handled automatically]
```

          direction TB
          D(Visit https://docs.example.com)

>           D-->E(fa:fa-lock SSL Certificate issued<br>dynamically)

Reviewer: ```suggestion
          D-->E(fa:fa-lock Correct SSL Certificate <br>automatically used)
```

Not sure what dynamically means here, but it's not really correct, since it's generating every 30 days?

>           E-->F(fa:fa-check Read the Docs matches<br>docs.example.com with<br>your project<br>)

Reviewer: This is somewhat confusing...

```suggestion
          E-->F(fa:fa-check Read the Docs knows<br> to serve your project <br>at ``docs.example.com``)
```

        end

        C-->automatic

Your documentation can have multiple secondary domains but only one **canonical** domain name.
Additional domains or subdomains will redirect to the canonical domain.

To make this work, Read the Docs generates a special text that you are responsible for copy-pasting to your domain's DNS.
In most cases, the ``CNAME`` record is used.
This is all that's needed for a web browser to resolve your domain name to Read the Docs' servers and for our servers to match the right documentation project.
You can find step-by-step instructions for this in :doc:`/guides/custom-domains`.

Read the Docs uses a :ref:`hosting:Content Delivery Network (CDN)` to host and serve your documentation pages.
This final step isn't changed by a custom domain
and therefore the response times are unaffected as the delivery of resources happens through the same CDN setup.

Considerations for custom domain usage
--------------------------------------

> Some Open Source projects have seen their domains expire.

Reviewer: We should put this in a style guide, as well..

For instance, this could be ``docs.example.com``.
This is great for maintaining a consistent brand for your documentation and application.

*By default*, your documentation is served from a Read the Docs :ref:`subdomain <hosting:subdomain support>` using the project's :term:`slug`:

* ``<slug>.readthedocs.io`` for |org_brand|
* ``<slug>.readthedocs-hosted.com`` for |com_brand|.

For example if you import your project and it gets the :term:`slug` ``example-docs``, it will be served from ``https://example-docs.readthedocs.io``.

How does it work?
-----------------

To begin using a custom domain, you will need to create a new custom domain for your project and add a DNS record for your new custom domain.
Once the DNS record has propagated, an SSL certificate will be automatically issued for your custom domain and your documentation will be configured to serve using the new domain and certificate.

In case you change your domain name, your documentation can have multiple secondary domains but only one canonical domain name.
Additional domains or subdomains will redirect to the canonical domain.

What to consider

> ----------------

Author: This is a new section, it's written as a reaction to a couple of unfortunate examples.

Reviewer: I don't love this heading, it's very ambiguous, but I don't have a strong suggestion for improving it.

Author: What about something like "Considerations for choosing a custom domain" ?

> ==============

Reviewer: Overall this page is a good upgrade, but I do feel like we're talking a lot about canonical domains, but not in a super clear way. We should probably have something like a Canonical domain section, and a bullet list of what it enables, which I think would be clearer?

Author: We also have this open - do we need to solve it in order to reference it?

![image](https://user-images.githubusercontent.com/374612/204396976-18f7e9cd-7772-4889-aae3-d7f87a25b8cc.png)

https://github.com/readthedocs/readthedocs.org/issues/9746

Reviewer: I don't quite follow? If our goal is to talk about canonical URLs here so much, then it does seem to have some overlap with those pages, so refactoring them in the same PR would make sense.

Custom domains allow you to serve your documentation from your own domain.
For instance, this could be ``docs.example.com``.
This is great for maintaining a consistent brand for your documentation and application.

*By default*, your documentation is served from a Read the Docs :ref:`subdomain <hosting:subdomain support>` using the project's :term:`slug`:

* ``<slug>.readthedocs.io`` for |org_brand|
* ``<slug>.readthedocs-hosted.com`` for |com_brand|.

> How does it work?
> -----------------

Reviewer: ```suggestion
How do custom domains work?
--------------------------------
```

We should generally avoid using `it` in headings,

Author: Good catch, especially with `sphinx.ext.autosectionlabel`, it seems important.

> To use a custom domain, you enter the domain in your Read the Docs project's administration and update your DNS provider with a new DNS entry.

Reviewer: I think we normally say `:guilabel:Admin` or `Dashboard`: https://docs.readthedocs.io/en/stable/glossary.html#term-dashboard -- but we should standardize on it.

Author: I opted for the "Dashboard" term -- the :guilabel: might make the "genre" blurry.

Reviewer 2: Dashboard and Admin are two different things.

The Dashboard is the "whole readthedocs.org application when the user is logged in". The Admin is the "project's admin page"

In this case, it seems the correct would be `` :guilabel:`Admin` `` --since the user has to add this domain in the admin of the project.

Reviewer 2: Actually, if you go to the current definition of it, it does not makes sense in this context:

![Screenshot_2022-11-30_19-37-02](https://user-images.githubusercontent.com/244656/204880919-d9305590-be7a-49bb-9a48-8bc9e9ea6163.png)

Author: Hmm, I see your point @humitos - I have misunderstood what @ericholscher was saying as two equal alternatives.

What I don't like so much is to put `:guilabel:` in the explanation because it indicates a concrete instruction. That's a widget we would normally put in a howto. In an explanation, it's distracting.

Is it possible that we add "project administration" or "project admin" to the glossary?

Author: Btw I really like to see a `:guilabel:`.. but in a How-To, Reference or Tutorial :) I'm afraid that it's a gateway drug or slippery slope for an explanation to end up doing how-to.

Reviewer: I think telling people *where* things are is definitely explanation, and well-suited for `:guilabel:`

Author: Alright, it's a good case to close a comment thread :100: I don't mind it really being `:guilabel:`, so let's do that and we're definitely gonna have a good documentation article :+1:

Reviewer 2: We have been using `:guilabel:` when we are talking directly about a link/tab/something in particular from the UI. When we are talking about a "general concept" we could omit the `:guilabel:`. We can change that behavior if we consider it's not accurate, but probably is another conversation 😄

> To use a custom domain, you enter the domain in your Read the Docs project's :term:`Dashboard` and update your DNS provider with a new DNS entry.

Author: ```suggestion
To use a custom domain, you enter the domain in your Read the Docs project's :guilabel:`Admin` and update your DNS provider with a new DNS entry.
```

> These two actions are all that are needed. Once the DNS record has propagated, Read the Docs automatically issues an SSL certificate through Cloudflare and starts serving your documentation.

Reviewer: Would a screenshot help here? Or a small workflow diagram or something? I find this section pretty confusing to follow.

Author: A diagram could be nice.

> In case you change your domain name, your documentation can have multiple secondary domains but only one **canonical** domain name.

Reviewer: ```suggestion
Your documentation can have multiple secondary domains but only one **canonical** domain name.
```

Additional domains or subdomains will redirect to the canonical domain.

To make this work, Read the Docs generates a special text that you are responsible for copy-pasting to your domain's DNS.
In most cases, the ``CNAME`` record is used.
This is all that's needed for a web browser to resolve your domain name to Read the Docs' servers and for our servers to match the right documentation project.

> You will find all the details in :doc:`/guides/custom-domains`.

Reviewer: ```suggestion
You can find step-by-step instructions for this in :doc:`/guides/custom-domains`.
```

What to consider
----------------

Some Open Source projects have seen their domains expire. Even prominent ones.

> It's important that you allocate the responsibility of your domain to a reliable actor in your organization.

Reviewer: ```suggestion
**It's important that you give the responsibility for managing your domain to someone reliable in your organization.**
```

> The domain will feature in search indexing and if you lose the domain, domain sharks may exploit this.

Reviewer: Nobody knows what a domain shark is..

```suggestion
Your canonical domain will feature in search indexing, so if you lose the domain your users could be sent to a page owned by someone else.
```

> You can choose if the domain is just an alias.

Reviewer: We should explain how to do this, or remove it... I think we're saying the same thing as above?

Author: I swapped it around. "canonical domain" was also used before it got explained and point about search indexing was repeated :+1:

> The **canonical** domain feature allows you to have several domains and choose which one search engines should respect when indexing your documentation project. The domain that you choose as your canonical domain is by far the most important one.

Reviewer: ```suggestion
The **canonical domain** feature allows you to have several domains and the canonical domain will be indexed by search engines.
The domain that you choose as your canonical domain is by far the most important one.
```

Custom domains allow you to serve your documentation from your own domain.
This is great for maintaining a consistent brand for your documentation and application.

By default, your documentation is served from a Read the Docs :ref:`subdomain <hosting:Subdomain support>` using the project's :term:`slug`:

* ``<slug>.readthedocs.io`` for |org_brand|
* ``<slug>.readthedocs-hosted.com`` for |com_brand|.

For example if you import your project and it gets the :term:`slug` ``example-docs``, it will be served from ``https://example-docs.readthedocs.io``.

How does it work?
-----------------

If you own a domain and would like to use it for your documentation, Read the Docs requires only :ref:`a single edit <adding_domain>` to the DNS settings of the domain.
After this, your documentation will be served from your domain and SSL will also be configured automatically.

In case you change your domain name, your documentation can have multiple secondary domains but only one canonical domain name.

> Additional domains or subdomains will redirect to the canonical domain.

Author: This section is new.

Reviewer 3: This section is a great addition. This is content that we've been missing for a while, and we don't always explain to the user (in terms the user is familiar with) what we actually do.

Author: Great! I'm glad that we could do the split and only have to write 1 new subsection, even though there's a lot of review feedback.

I hope I can present some PRs that invite a faster process after this.

For example if you import your project and it gets the :term:`slug` ``example-docs``, it will be served from ``https://example-docs.readthedocs.io``.

How does it work?
-----------------

> If you own a domain and would like to use it for your documentation, Read the Docs requires only :ref:`a single edit <adding_domain>` to the DNS settings of the domain.
> After this, your documentation will be served from your domain and SSL will also be configured automatically.

Reviewer 3: This voice style is a great fit for marketing content on our site and communications, where we're trying to get the reader's attention and our focus isn't on technically precise language.

However, in documentation, this voice tends to hide technical details in marketing copy. Phrasing in the vein of "it's so simple" is good to avoid, as we can't guarantee this claim or that the user's experience matches this. If the reader is on this documentation page, chances are they are having a different experience with the feature.

I'd suggest aiming for clearer language and explicitly call out the process required more, something closer to:

```suggestion
To begin using a custom domain, you will need to create a new custom domain for your project and add a DNS record for your new custom domain.
Once the DNS record has propagated, an SSL certificate will be automatically issued for your custom domain and your documentation will be configured to serve using the new domain and certificate.
```

In case you change your domain name, your documentation can have multiple secondary domains but only one canonical domain name.
Additional domains or subdomains will redirect to the canonical domain.

.. seealso::

>     Eager to get started? Have a look at :doc:`/guides/custom-domains`

Reviewer 3: This is another good example of copy that fits really well on our website, but makes for usability issues in our docs. If we're expecting the user to click through to these docs, we should tell the more explicitly what they'll find -- this doesn't communicate that yet.

I like the pattern from some other pages where we use a definition list in a seealso. It gives a chance to display the full page title without having to rework the page title to fit a sentence, and it gives a place to describe why the link is relevant to the user:

```rst
.. seealso::

    :doc:`/guides/custom-domains`
        Information on creating and managing custom domains, and common configurations you might use to set up your domain
```

Author: The "How to ABC" title is great for this admonition. I wonder if we can also fit the 1-line description into the meta data for the page. For instance if we wrote a role for accessing this: https://www.sphinx-doc.org/en/master/extdev/envapi.html#sphinx.environment.BuildEnvironment.metadata

.. seealso::

    :doc:`/guides/custom-domains`
        Information on creating and managing custom domains,

>         and common configurations you might use to set up your domain

Reviewer: ❤️

The domain that you choose as your canonical domain is by far the most important one.
If you lose the canonical domain, someone else can set up a website that search results will end up referring to.

Canonical URLs
--------------

A `canonical URL`_
allows you to specify the preferred version of a web page to prevent duplicated content.
They are mainly used by search engines to link users to the correct
version and domain of your documentation.

If canonical URL's aren't used,
it's easy for outdated documentation to be the top search result for various pages in your documentation.
This is not a perfect solution for this problem,
but generally people finding outdated documentation is a big problem,
and this is one of the suggested ways to solve it from search engines.

.. _canonical URL: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls

How Read the Docs generates canonical URLs
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The canonical URL takes into account:

* The default version of your project (usually "latest" or "stable").

> * The canonical :doc:`custom domain </custom-domains>` if you have one,

Author: Yes, absolutely, the only reason to copy in the contents was for "Sketching" as the commit message went... so just taking a look and getting some ideas :)
